client/my-sites/controller.js

```diff
@@ -548,11 +548,26 @@ export function siteSelection( context, next ) {
 	const isUnlinkedCheckoutNotBoost =
 		isUnlinkedCheckout && ! context.pathname.includes( 'jetpack_boost' );
 
+	const currentPlanSlug = getSitePlanSlug( getState(), siteId );
+	const shouldUpdateStateAfterUpgrade =
+		context.pathname.includes( 'plans/my-plan/trial-upgraded/' ) &&
+		[ PLAN_FREE, PLAN_JETPACK_FREE ].includes( currentPlanSlug );
+
 	if ( siteId && ! isUnlinkedCheckoutNotBoost ) {
 		// onSelectedSiteAvailable might render an error page about domain-only sites or redirect
 		// to wp-admin. In that case, don't continue handling the route.
 		dispatch( setSelectedSiteId( siteId ) );
-		if ( onSelectedSiteAvailable( context ) ) {
+
+		// This will fetch the site and update the state after the plan is upgraded
+		if ( shouldUpdateStateAfterUpgrade ) {
+			dispatch( requestSite( siteFragment ) )
+				.catch( () => null )
+				.then( () => {
+					if ( onSelectedSiteAvailable( context ) ) {
+						next();
+					}
+				} );
+		} else if ( onSelectedSiteAvailable( context ) ) {
```

Comment:
This adds a bit of duplication, which indicates a potential code smell. Could we improve so we don't call the `onSelectedSiteAvailable( context )` condition twice?